docs/proposals/scalable-alertmanager.md

```diff
@@ -0,0 +1,78 @@
+---
+title: "Scalable Alertmanager"
+linkTitle: "Scalable Alertmanager"
+weight: 1
+slug: scalable-alertmanager
+---
+
+- Author: [Josh Abreu](https://github.com/gotjosh)
+- Date: December 2020
+- Status: Proposed
+
+## Context and Background
+
+The Cortex Alertmanager at its current state supports high-availability by using the same technique as the upstream Prometheus Alertmanager: we gossip silences and notifications between replicas to achieve eventual consistency. This allows us to tolerate machine failure without interruption of service. The caveat here is traffic between the ruler and Alertmanagers must not be load balanced as alerts themselves are not gossiped.
+
+By itself it is not horizontally scalable; it is recommended to run a maximum of 3 replicas (for high availability). Each alertmanager replica will contain an internal instance for every tenant. The Alertmanager uses roughly ~3.7MB of memory and ~0.001 CPU cores per tenant, with an average tenant having ~10 active alerts and ~1 silences. These stats were measured from a cluster with ~80 QPS (Alerts received Per Second), ~40 NPS (Notifications Per Second), and ~700 configurations over a single replica.
+
+
+## Problem and Requirements
+
+Current numbers show a reasonably sized machine can handle 2000 tenants in the current service. We would like to be able to scale this up to 10x without increasing the machine size; i.e. we would like to make Cortex Alertmanager service horizontally scalable.
+
+Furthermore, we desire to preserve the following characteristics:
+
+- A single Cortex Alertmanager replica crashing or exiting abruptly should not cause externally-visible downtime or failure to deliver notifications.
+- Users should have an eventually consistent view of all the alerts current firing and under active notification, favouring availability of the Cortex Alertmanager over consistency.
+- We should be able to scale up, down and roll out new versions of the service without any service interruption or data loss.
+
+## Design
+
+This is a big design, and as such we have divided the problem up into 4 areas: Routing & Sharding, Persistence & State, Replication & Consistency, and Architecture. The options and solutions I propose are divided along these lines to aid clarity.
+
+### Routing & Sharding
+
+To achieve horizontal scalability, we need to distribute the workload among replicas of the service. We need to choose an appropriate field to use to distribute the workload. The field must be present on all the API requests to the Alertmanager service.
+
+**We propose the sharding on Tenant ID**. The simplicity of this implementation, would allow us to get up and running relatively quickly whilst helping us validate assumptions. We intend to use the existing ring code to manage this. Other options such as tenant ID + receiver or Tenant ID + route are relatively complex as distributor components (in this case the Ruler) would need to be aware of Alertmanager configuration.
```

Comment:
I think this con is fine for the time being If the current design can support ~2000 average tenants, it could with a little handholding support a tenant 2000 times normal on by fiddling with sharding overrides, so I don't think this limitation is reason enough to not choose the simple option 1, shard on tenant id.